bundles/org.openhab.core/src/main/java/org/openhab/core/util/ColorUtil.java

```diff
@@ -421,82 +437,110 @@ public static HSBType xyToHsb(double[] xy) throws IllegalArgumentException {
-            g -= min;
-            b -= min;
+            for (int i = 0; i < rgb.length; i++) {
+                rgb[i] -= min;
+            }
         }
 
         // rescale
-        double max = Math.max(r, Math.max(g, b));
+        double max = Math.max(rgb[0], Math.max(rgb[1], rgb[2]));
         if (max > 1.0) {
-            r /= max;
-            g /= max;
-            b /= max;
+            for (int i = 0; i < rgb.length; i++) {
+                rgb[i] /= max;
+            }
         }
 
-        r = compand(r);
-        g = compand(g);
-        b = compand(b);
+        // remove gamma correction
+        for (int i = 0; i < rgb.length; i++) {
+            rgb[i] = rgb[i] <= 0.0031308 ? 12.92 * rgb[i] : (1.0 + 0.055) * Math.pow(rgb[i], (1.0 / 2.4)) - 0.055;
+        }
 
         // rescale
-        max = Math.max(r, Math.max(g, b));
+        max = Math.max(rgb[0], Math.max(rgb[1], rgb[2]));
         if (max > 1.0) {
-            r /= max;
-            g /= max;
-            b /= max;
+            for (int i = 0; i < rgb.length; i++) {
+                rgb[i] /= max;
+            }
         }
 
-        LOGGER.trace("xy: {} - XYZ: {} {} {} - RGB: {} {} {}", xy, X, Y, Z, r, g, b);
+        // convert double[] to PercentType[]
+        PercentType[] rgbPercents = new PercentType[rgb.length];
+        for (int i = 0; i < rgb.length; i++) {
+            rgbPercents[i] = new PercentType(new BigDecimal(rgb[i]).multiply(BIG_DECIMAL_100, COLOR_MATH_CONTEXT));
+        }
 
-        return rgbToHsb(new PercentType[] { convertDoubleToColorPercent(r), convertDoubleToColorPercent(g),
-                convertDoubleToColorPercent(b) });
+        HSBType hsb = rgbToHsb(rgbPercents);
+
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace("{}", String.format(
+                    "xyY:[%.6f,%.6f,%.6f] - xyz:[%.6f,%.6f,%.6f] - XYZ:[%.6f,%.6f,%.6f] - RGB':[%.6f,%.6f,%.6f] - RGB:[%.6f,%.6f,%.6f] - HSB:[%.6f,%.6f,%.6f] (xyForced:%b)",
+                    xyY[0], xyY[1], Y, x, y, z, X, Y, Z, rgbPrime[0], rgbPrime[1], rgbPrime[2], rgb[0], rgb[1], rgb[2],
+                    hsb.getHue().doubleValue(), hsb.getSaturation().doubleValue(), hsb.getBrightness().doubleValue(),
+                    xy.x != xyY[0] || xy.y != xyY[1]));
+        }
+
+        return hsb;
     }
 
     /**
-     * Gamma correction (inverse sRGB companding)
-     *
-     * @param value the value to process
-     * @return the processed value
+     * Get an array of int from an array of PercentType.
      */
-    private static double inverseCompand(double value) {
-        return value > 0.04045 ? Math.pow((value + 0.055) / (1.0 + 0.055), 2.4) : value / 12.92;
+    private static int[] getIntArray(PercentType[] percents) {
+        int[] ints = new int[percents.length];
+        for (int i = 0; i < percents.length; i++) {
+            ints[i] = percents[i].toBigDecimal().multiply(BIG_DECIMAL_255)
+                    .divide(BIG_DECIMAL_100, 0, RoundingMode.HALF_UP).intValue();
```

Comment:
Use a constant `BIG_DECIMAL_2_POINT_55` here, that save one operation per conversion.

Reply:
Done.

Reply:
@J-N-K I am reverting this change as it breaks the JUnit tests. Not only do the tests fail, but the actual conversions are self evidently wrong too.